controllers/keystoneapi_controller.go

```diff
@@ -109,6 +109,25 @@ func (r *KeystoneAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, err
 	}
 
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		// Register the finalizer immediately to avoid orphaning resources on delete
+		err := r.Update(ctx, instance)
+
+		return ctrl.Result{}, err
```

Comment:
shouldn't we requeue here as well?

Reply:
I think in these cases that we don't need to explicitly include `Requeue: true` because the `Update()` of the instance will already have caused the instance to be requeued

Reply:
when we update instance with added finalizer to be set, the next reconcile from a queue could still not see it commited by kube-apisserver, right? So we would add it here once again, call update again, and the story repeats with a busy loop again and again. To prevent that, we probably should not add a finalizer unconditionally *but which condition to use, if all we have is cached view into the system state*?

Reply:
Is there a way to get information about the context this reconcile has been called with? Like requed by us earlier, or auto-requed after Update() or Patch() call, or triggered by watcher? Having that info might help to aboid the busy loop I described above.

But it shouldn't happen, if there is guarantees that informer/watcher do not place a reconcile event for added finalizer until kube-apiserver applies it and it pops up updated in the cache.

Reply:
@mdbooth WDYT ^^?

Reply:
> As soon as you call Update() here the watch is going to update the shared informer, and this reconciler is queued to be called again with the state of the object at this point.

So if I get that right, the queued reconcile for this Update() call that adds a finalizer, is guaranteed to have that finalizer in the cache by the time we hit this code again.

still not sure if we should be adding finalizer unconditionally, as there could be later reconcile loops unrelated to that Update() for finalizer? And then, we should prefer to not add it again and avoid generating unnecessary updates in the queue?..

Reply:
Just to respond to one point specifically, adding the finalizer as we are doing here is no unconditional.  The docs [1] state:

```
AddFinalizer accepts an Object and adds the provided finalizer if not present. It returns an indication of whether it updated the object's list of finalizers. 
```

The second sentence leads me to believe that `AddFinalizer` will only return true if the instance indeed lacked the finalizer and thus needed it to be added.

[1] https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#AddFinalizer

Reply:
I agree that update on ourselves will trigger a new reconcile so no requeue is needed.